datafusion/functions/src/math/log.rs

```diff
@@ -0,0 +1,251 @@
+        }
+        // note in f64::log params order is different than in sql. e.g in sql log(base, x) == f64::log(x, base)
+        let arr: ArrayRef = match args[0].data_type() {
+            DataType::Float64 => match base {
+                ColumnarValue::Scalar(ScalarValue::Float32(Some(base))) => {
+                    Arc::new(make_function_scalar_inputs!(x, "x", Float64Array, {
+                        |value: f64| f64::log(value, base as f64)
+                    }))
+                }
+                ColumnarValue::Array(base) => Arc::new(make_function_inputs2!(
+                    x,
+                    base,
+                    "x",
+                    "base",
+                    Float64Array,
+                    { f64::log }
+                )),
+                _ => {
+                    return exec_err!("log function requires a scalar or array for base")
+                }
+            },
+
+            DataType::Float32 => match base {
+                ColumnarValue::Scalar(ScalarValue::Float32(Some(base))) => {
+                    Arc::new(make_function_scalar_inputs!(x, "x", Float32Array, {
+                        |value: f32| f32::log(value, base)
+                    }))
+                }
+                ColumnarValue::Array(base) => Arc::new(make_function_inputs2!(
+                    x,
+                    base,
+                    "x",
+                    "base",
+                    Float32Array,
+                    { f32::log }
+                )),
+                _ => {
+                    return exec_err!("log function requires a scalar or array for base")
+                }
+            },
+            other => {
+                return exec_err!("Unsupported data type {other:?} for function log")
+            }
+        };
+
+        Ok(ColumnarValue::Array(arr))
+    }
+
+    /// Simplify the `log` function by the relevant rules:
+    /// 1. Log(a, 1) ===> 0
+    /// 2. Log(a, Power(a, b)) ===> b
+    /// 3. Log(a, a) ===> 1
+    fn simplify(
+        &self,
+        args: Vec<Expr>,
+        info: &dyn SimplifyInfo,
+    ) -> Result<ExprSimplifyResult> {
+        let mut number = &args[0];
+        let mut base =
+            &Expr::Literal(ScalarValue::new_ten(&info.get_data_type(number)?)?);
+        if args.len() == 2 {
+            base = &args[0];
+            number = &args[1];
+        }
+
+        match number {
+            Expr::Literal(value)
+                if value == &ScalarValue::new_one(&info.get_data_type(number)?)? =>
+            {
+                Ok(ExprSimplifyResult::Simplified(Expr::Literal(
+                    ScalarValue::new_zero(&info.get_data_type(base)?)?,
+                )))
+            }
+            // todo: match type instead of name here
+            Expr::ScalarFunction(ScalarFunction {
+                func_def: ScalarFunctionDefinition::UDF(fun),
+                args,
+            }) if base == &args[0] && fun.name() == "power" => {
```

Comment:
I wasn't able to figure out how to match the exact `PowerFunc` type here. Same for a rule in `power.rs`.

Reply:
I think you can do something like 

```rust
let is_power = fun.as_any().downcast_ref::<PowerUDF>().is_some();
```

If that doesn't work I can try and look more into it on Monday

Reply:
This worked with a little modification. Thank you!

```rust
fun.as_ref().inner().as_any().downcast_ref::<PowerFunc>().is_some()
```

Reply:
I took a look over the normal things that catch me (monotonicity, aliases, tests) and nothing stood out - LGTM